lm75bd/lm75bd.c

```diff
@@ -25,9 +26,46 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
   return ERR_CODE_SUCCESS;
 }
 
+#define TEMP_REG_WRITE_BUFF_SIZE 1U
+#define TEMP_READ_BUFF_SIZE 2U
 error_code_t readTempLM75BD(uint8_t devAddr, float *temp) {
   /* Implement this driver function */
-  
+  if (temp == NULL) {
+    return ERR_CODE_INVALID_ARG;
+  }
+
+  error_code_t errCode;
+
+  // Set up the buffer to store the temperature register address
+  // 0: Register address
+  uint8_t writeBuff[TEMP_REG_WRITE_BUFF_SIZE] = { LM75BD_REG_TEMP };
+
+  // Send the pointer to set up reading from the temperature register
+  errCode = i2cSendTo(devAddr, writeBuff, TEMP_REG_WRITE_BUFF_SIZE);
+  RETURN_IF_ERROR_CODE(errCode);
+
+  // Set up the buffer to store the temperature data received
+  uint8_t readBuff[TEMP_READ_BUFF_SIZE] = { 0, 0 };
+
+  // Read from the temperature register
+  errCode = i2cReceiveFrom(devAddr, readBuff, TEMP_READ_BUFF_SIZE);
+  RETURN_IF_ERROR_CODE(errCode);
+
+  // Convert temperature register value to degrees Celsius
+  // 1. Extract the temperature data (D10 to D0)
+  // | D10 D9 D8 D7 D6 D5 D4 D3 | D2 D1 D0 X X X X X |
+  int16_t temp_data = (readBuff[0] << 3) & 0x7F8;
+  temp_data |= (readBuff[1] >> 5) & 0x07;
+
+  // 2. Checking if D10 is 0 (positive) or 1 (negative)
+  // and sign extending to allow for two's complement if D10 is 1
+  if (temp_data & (1 << 10)) {
+    temp_data |= ~0x7FF;
+  }
```

Comment:
This code is needless complicated. You can just use ( (readBuff[0] << 8) | readBuff[1]) >> 5 instead